### ISSUE TITLE
Should global shap values depend on the variance of the feature?

### ISSUE THREAD

**Issue by jorgecarleitao** [shap/shap#696]
I am struggling to make sense of how "global feature importance" is computed, and why it should depend on the variance of a feature.

Specifically, in the source code, "global shap values" is computed with the formula: `np.abs(shap).mean(0)` (see [here](https://github.com/slundberg/shap/blob/6bf9df73180b621d8dd5be088c767d7bf0262e2a/shap/plots/summary.py#L382), which amounts to compute the average of the absolute value of the shap values).

Let's take the following example: It basically sets some random variables, relates them with y, fits a lasso model, and computes the respective shap values and global shap values.

```
import numpy as np
from shap import LinearExplainer
from sklearn.linear_model import LassoCV


np.random.seed(1)

a1 = np.random.normal(size=500, scale=1)
a2 = np.random.normal(size=500, scale=1)
a3 = np.random.normal(size=500, scale=1)

x = np.array([a1, a2, a3]).T
y = 100 * a1 + 20 * a2 + np.random.normal(size=500, scale=0.01)

model = LassoCV(cv=5)

model.fit(x, y)

explainer = LinearExplainer(model, x)
shap = explainer.shap_values(x)

values = np.abs(shap).mean(0)
print(values)
# [78.9061823  15.19658609  0.        ]
```

The issue for me is that these values depend on the variance of `a{i}`.

Why? `mean(abs(shap))` will on average be equal to `np.sqrt(2/np.pi) * np.std(shap, axis=0)` (see [expectation of half-normal distribution](https://en.wikipedia.org/wiki/Half-normal_distribution)). In this case, since we are in the linear models and `a{i}` is centered, we get that `std(shap) = std((x - m(x)) * c) = std(x) * c`. We can confirm this by also printing

     print(np.sqrt(2/np.pi) * np.std(shap, axis=0))
     print(np.sqrt(2/np.pi) * np.std(x, axis=0) * model.coef_)

Therefore, changing the initial scale of, e.g. `a2`, we get a different importance. Indeed, changing the line above to 

     a2 = np.random.normal(size=500, scale=10)

leads to the output

     [ 77.43738977 152.6175833    0.        ]

i.e. `a2` is now more important than `a1`, even though we only increased its variance. Shouldn't the global feature importance be independent of the variance of the feature?

**Comment by ivan-marroquin** [comment]
Hi all,

Thanks @jorgecarleitao  for bringing this point. I have the same concerns. Another drawback of using the mean is to get a result of 0 (just because there is the same amount of positive and negative shap values). 

Would it not be better to use another metric to measure the average for shap values? What about median or median absolute deviation?

Kind regards,
Ivan

**Comment by slundberg** [comment]
@jorgecarleitao sorry I missed this over the summer. You are right that the global importance of a feature in a linear model will depend on the variance of the inputs feature. The reason this is right is because the model itself is sensitive to the variance of the input feature. To see this, imagine increasing the variance of a single input feature for a linear model to a huge value. As long as the coefficient for that input is not zero you will eventually get to a point when the model's output is entirely driven by that one feature. Conversely imagine reducing the variance of all the other features to nearly zero. Now the model will effectively not depend on the values of those features. The summary is that depending on the variance is correct and a feature of measuring global feature importance with SHAP (in my opinion).

@ivan-marroquin I am not sure I understand your additional concern, as long as we take the absolute value we will not have positive and negative SHAP values canceling each other out.

**Comment by ivan-marroquin** [comment]
Hi @slundberg ,

Thanks for your quick answer. Indeed, the use of the absolute takes care of the concern that I raised. But, I still have this issue. I believe that positive and negative shap values are important, and they should be used in concert to derive shap values. 

The absolute value collapses all that information and into positive values. It seems to me that the average is no longer representative of the original data.

So, if we compute the median from the original shap values and because the median is more robust than the mean, it could provide a better central tendency measure. Thereafter, we take the absolute value assuming that we are only interest in the strength of the average. 

Ivan

**Comment by slundberg** [comment]
Good questions! Positive and negative values are important, but if you allow them to cancel each other out when measuring global feature importance then you will not be measuring the importance of the features for the model, but rather how different the impact of each feature is between the samples you are explaining and your background set of samples. For example if you run `shap.TreeExplainer(model, X).shap_values(X).mean(0)` you will get an all-zero vector.

By the way, I think that the "right" way to get global feature importance is to explain the loss of the model and then just take the mean. By explaining the loss you avoid the need for any absolute operators.

**Comment by ivan-marroquin** [comment]
Hi @slundberg ,

Thanks for your comments. Are you planing to test the use of the loss to get global feature importance? 

Ivan

**Comment by slundberg** [comment]
We briefly tested it in https://rdcu.be/b0z70 (supplementary feature selection experiments) and will do more work on it.

**Comment by Corne173** [comment]
If you look at the SHAP source code for generating global feature importance it take the absolute value of the attributions and then surmises them.

**Comment by ivan-marroquin** [comment]
Hi,

Sorry for my late reply.

Since SHAP values can be positive and negative values. Then, I believe using the mean of the absolute values may not be representative of the original SHAP values. As suggested by @slundberg, the median (without using absolute values) provides a better measure of central tendency.

Hope this helps,
Ivan
